### ISSUE TITLE
Piano: Don't pass SB16 device FD to WindowServer

### ISSUE THREAD

**Issue by gunnarbeutner** [SerenityOS/serenity#7641]
For some reason (which still eludes me) Piano manages to pass its SB16 file descriptor to WindowServer in a `set_window_backing_store` IPC call:

```
courage:/home/anon # lsof -p 31
COMMAND                       PID PGID USER         FD NAME
Piano                          31   31 anon          0 /dev/pts/0
Piano                          31   31 anon          1 /dev/pts/0
Piano                          31   31 anon          2 /dev/pts/0
Piano                          31   31 anon          3 fifo:21
Piano                          31   31 anon          4 fifo:21
Piano                          31   31 anon          5 socket:/tmp/portal/window (connected to pid 23)
Piano                          31   31 anon          6 :anonymous-file:
Piano                          31   31 anon          7 socket:/tmp/portal/clipboard (connected to pid 11)
Piano                          31   31 anon          8 socket:/tmp/portal/audio (connected to pid 18)
Piano                          31   31 anon          9 device:42,42 (SB16)
Piano                          31   31 anon         10 /res/fonts/LiberationSerif-Bold.ttf
Piano                          31   31 anon         11 /res/fonts/LiberationSerif-BoldItalic.ttf
Piano                          31   31 anon         12 /res/fonts/LiberationSerif-Italic.ttf
Piano                          31   31 anon         13 /res/fonts/LiberationSerif-Regular.ttf
Piano                          31   31 anon         14 /res/fonts/SerenitySans-Regular.ttf
courage:/home/anon # 
```

```
courage:/home/anon # lsof -p 23
COMMAND                       PID PGID USER         FD NAME
WindowServer                   23    0 window        0 device:1,3 (NullDevice)
WindowServer                   23    0 window        1 device:1,3 (NullDevice)
WindowServer                   23    0 window        2 device:1,3 (NullDevice)
WindowServer                   23    0 window        3 :anonymous-file:
WindowServer                   23    0 window        4 fifo:15
WindowServer                   23    0 window        5 fifo:15
WindowServer                   23    0 window        6 device:85,0 (KeyboardDevice)
WindowServer                   23    0 window        7 device:10,0 (MouseDevice)
WindowServer                   23    0 window        8 device:29,0 (FramebufferDevice)
WindowServer                   23    0 window        9 socket:/tmp/portal/window (accepted from pid 16)
WindowServer                   23    0 window       10 /res/fonts/LiberationSerif-Bold.ttf
WindowServer                   23    0 window       11 /res/fonts/LiberationSerif-BoldItalic.ttf
WindowServer                   23    0 window       12 /res/fonts/LiberationSerif-Italic.ttf
WindowServer                   23    0 window       13 /res/fonts/LiberationSerif-Regular.ttf
WindowServer                   23    0 window       14 /res/fonts/SerenitySans-Regular.ttf
WindowServer                   23    0 window       15 socket:/tmp/portal/window (accepted from pid 22)
WindowServer                   23    0 window       16 socket:/tmp/portal/window (accepted from pid 24)
WindowServer                   23    0 window       17 socket:/tmp/portal/window (accepted from pid 10)
WindowServer                   23    0 window       18 socket:/tmp/portal/window (accepted from pid 15)
WindowServer                   23    0 window       19 socket:/tmp/portal/window (accepted from pid 17)
WindowServer                   23    0 window       20 socket:/tmp/portal/window (listening)
WindowServer                   23    0 window       21 socket:/tmp/portal/wm (listening)
WindowServer                   23    0 window       22 socket:/tmp/portal/window (accepted from pid 12)
WindowServer                   23    0 window       23 socket:/tmp/portal/wm (accepted from pid 24)
WindowServer                   23    0 window       24 :anonymous-file:
WindowServer                   23    0 window       25 :anonymous-file:
WindowServer                   23    0 window       26 :anonymous-file:
WindowServer                   23    0 window       27 :anonymous-file:
WindowServer                   23    0 window       28 :anonymous-file:
WindowServer                   23    0 window       29 :anonymous-file:
WindowServer                   23    0 window       30 :anonymous-file:
WindowServer                   23    0 window       31 :anonymous-file:
WindowServer                   23    0 window       32 :anonymous-file:
WindowServer                   23    0 window       33 :anonymous-file:
WindowServer                   23    0 window       34 :anonymous-file:
WindowServer                   23    0 window       35 :anonymous-file:
WindowServer                   23    0 window       36 :anonymous-file:
WindowServer                   23    0 window       37 :anonymous-file:
WindowServer                   23    0 window       38 :anonymous-file:
WindowServer                   23    0 window       39 :anonymous-file:
WindowServer                   23    0 window       40 :anonymous-file:
WindowServer                   23    0 window       41 :anonymous-file:
WindowServer                   23    0 window       42 :anonymous-file:
WindowServer                   23    0 window       43 :anonymous-file:
WindowServer                   23    0 window       44 :anonymous-file:
WindowServer                   23    0 window       45 :anonymous-file:
WindowServer                   23    0 window       46 :anonymous-file:
WindowServer                   23    0 window       47 :anonymous-file:
WindowServer                   23    0 window       48 :anonymous-file:
WindowServer                   23    0 window       49 :anonymous-file:
WindowServer                   23    0 window       50 :anonymous-file:
WindowServer                   23    0 window       51 :anonymous-file:
WindowServer                   23    0 window       52 :anonymous-file:
WindowServer                   23    0 window       53 :anonymous-file:
WindowServer                   23    0 window       54 :anonymous-file:
WindowServer                   23    0 window       55 :anonymous-file:
WindowServer                   23    0 window       56 :anonymous-file:
WindowServer                   23    0 window       57 :anonymous-file:
WindowServer                   23    0 window       58 :anonymous-file:
WindowServer                   23    0 window       59 :anonymous-file:
WindowServer                   23    0 window       60 :anonymous-file:
WindowServer                   23    0 window       61 :anonymous-file:
WindowServer                   23    0 window       62 :anonymous-file:
WindowServer                   23    0 window       63 :anonymous-file:
WindowServer                   23    0 window       64 :anonymous-file:
WindowServer                   23    0 window       65 :anonymous-file:
WindowServer                   23    0 window       66 :anonymous-file:
WindowServer                   23    0 window       67 :anonymous-file:
WindowServer                   23    0 window       68 :anonymous-file:
WindowServer                   23    0 window       69 :anonymous-file:
WindowServer                   23    0 window       70 :anonymous-file:
WindowServer                   23    0 window       71 :anonymous-file:
WindowServer                   23    0 window       72 :anonymous-file:
WindowServer                   23    0 window       73 :anonymous-file:
WindowServer                   23    0 window       74 :anonymous-file:
WindowServer                   23    0 window       75 :anonymous-file:
WindowServer                   23    0 window       76 :anonymous-file:
WindowServer                   23    0 window       77 :anonymous-file:
WindowServer                   23    0 window       78 :anonymous-file:
WindowServer                   23    0 window       79 :anonymous-file:
WindowServer                   23    0 window       80 :anonymous-file:
WindowServer                   23    0 window       81 :anonymous-file:
WindowServer                   23    0 window       82 :anonymous-file:
WindowServer                   23    0 window       83 :anonymous-file:
WindowServer                   23    0 window       84 :anonymous-file:
WindowServer                   23    0 window       85 :anonymous-file:
WindowServer                   23    0 window       86 :anonymous-file:
WindowServer                   23    0 window       87 :anonymous-file:
WindowServer                   23    0 window       88 :anonymous-file:
WindowServer                   23    0 window       89 :anonymous-file:
WindowServer                   23    0 window       90 :anonymous-file:
WindowServer                   23    0 window       91 :anonymous-file:
WindowServer                   23    0 window       92 :anonymous-file:
WindowServer                   23    0 window       93 :anonymous-file:
WindowServer                   23    0 window       94 :anonymous-file:
WindowServer                   23    0 window       95 :anonymous-file:
WindowServer                   23    0 window       96 :anonymous-file:
WindowServer                   23    0 window       97 :anonymous-file:
WindowServer                   23    0 window       98 :anonymous-file:
WindowServer                   23    0 window       99 :anonymous-file:
WindowServer                   23    0 window      100 :anonymous-file:
WindowServer                   23    0 window      101 :anonymous-file:
WindowServer                   23    0 window      102 :anonymous-file:
WindowServer                   23    0 window      103 :anonymous-file:
WindowServer                   23    0 window      104 :anonymous-file:
WindowServer                   23    0 window      105 :anonymous-file:
WindowServer                   23    0 window      106 :anonymous-file:
WindowServer                   23    0 window      107 :anonymous-file:
WindowServer                   23    0 window      108 :anonymous-file:
WindowServer                   23    0 window      109 :anonymous-file:
WindowServer                   23    0 window      110 :anonymous-file:
WindowServer                   23    0 window      111 :anonymous-file:
WindowServer                   23    0 window      112 :anonymous-file:
WindowServer                   23    0 window      113 :anonymous-file:
WindowServer                   23    0 window      114 :anonymous-file:
WindowServer                   23    0 window      115 :anonymous-file:
WindowServer                   23    0 window      116 :anonymous-file:
WindowServer                   23    0 window      117 :anonymous-file:
WindowServer                   23    0 window      118 socket:/tmp/portal/window (accepted from pid 31)
WindowServer                   23    0 window      119 :anonymous-file:
WindowServer                   23    0 window      120 :anonymous-file:
WindowServer                   23    0 window      121 device:42,42 (SB16)
```

WindowServer then tries to `mmap` that FD which fails and causes WindowServer to crash.

**Comment by kleinesfilmroellchen** [comment]
This may be mitigated by simply making Piano not open /dev/audio at all, but using the (newer) AudioServer connection method, i.e. the proper way.

**Comment by gunnarbeutner** [comment]
Well, sure, but I think there's a different, underlying problem. It seems like it's just passing an entirely unrelated file descriptor to WindowServer there.

**Comment by kleinesfilmroellchen** [comment]
https://github.com/SerenityOS/serenity/blob/67a5e9f0184eac51a7451f540845e49dfdc0b34a/Userland/Applications/Piano/main.cpp#L52

**Comment by tomuta** [comment]
@awesomekling  I'm going to re-open this as we discovered that the supposed fix was merely just a half-baked cover-up after all...

**Comment by gunnarbeutner** [comment]
So is this (probably) fixed now that #8657 is merged?

**Comment by tomuta** [comment]
> So is this (probably) fixed now that #8657 is merged?

That would be my guess.
